### ISSUE TITLE
Automatically try casting arguments to Operators

### ISSUE THREAD

**Issue by adler-j** [odlgroup/odl#100]
Currently in the type checking of `__call__` we do

```python
if x not in self.range:
    raise
```

How about we instead try:

```python
if x not in self.range:
    x = self.range.element(x)
```

Thus if `x`is both not in the range, nor castable to a element in the range, we throw an error.

The main motivation would be twofold. First, users can do simple tests:

```python
>>> my_square_op([1, 2, 3])
Rn(3).element([1.0, 4.0, 9.0])
```

Further, users could simplify using product-space wrappers:

Compare
```python
>>> z = my_prod_space.element([x, y])
>>> op(z)
```

to

```python
>>> op([x, y])
```

**Comment by kohr-h** [comment]
I'm in favor of this idea, on the one hand because it simplifies usage, and on the other, since it's more pythonic. We could apply the same strategy on the `range` side, so one could implement the matrix multiplication operator by
```python
def _call(self, x):
    return self.matrix.dot(x)
```
instead of 
```python
def _call(self, x):
    return self.range.element(self.matrix.dot(x))
```
The latter wrapping could happen in the `__call__()` method, thus making it easier to write operators.

**Comment by adler-j** [comment]
Agree on that aswell, allows for some nice syntax.

**Comment by adler-j** [comment]
Is this being done as part of the vectorization branch, or can we do it separately?

**Comment by kohr-h** [comment]
At least I [added a comment](https://github.com/odlgroup/odl/blob/issue-57__operator/odl/space/ntuples.py#L145-L146) there. Since it's closely related, I can take care of that in the branch.

**Comment by adler-j** [comment]
I closed the broadcasting issue #103 yesterday, so the comment is "outdated", adding this should be simple, hardest part is probably to add some tests.

**Comment by kohr-h** [comment]
Alright, I'll merge it into my branch. We should probably keep this open then until there are tests for this new behavior.

**Comment by kohr-h** [comment]
Closed by #116
